### ISSUE TITLE
Paramaterisation of constructor arguments

### ISSUE THREAD

**Issue by MarkUfland** [AutoMapper/AutoMapper#2014]
I'm not sure if this is an issue but I can't seem to get it to work. I have an immutable class that I need to map to from another object. The class does not have an empty constructor and the properties have private sets.  There are arguments that I need to pass to the constructor at runtime so I want to use the resolution context to do that along the lines of 

cfg.CreateMap<Foo, Bar>()
   .ForCtorParam(dest => dest.Baz, opt => opt.ResolveUsing(res => res.Context.Options.Items["thing"]));

Mapper.Map<Foo, Bar>(foo, opt => opt.Items["thing"] = User.Name);

However, I can't seem to get hold of the resolution context in ForCtorParam only ForMember so the above does not work.  Is there a way to do this?

Thanks

**Comment by lbargaoanu** [comment]
I guess an  extra overload is needed. PR?

**Comment by MarkUfland** [comment]
I took a clone of the repo but the core project does not want to load in VS 2015. It seems to think the project file needs modifying for an older version of MSbuild.  Is there something I need to do?

**Comment by jbogard** [comment]
Install VS 2017.

**Comment by lbargaoanu** [comment]
You could try constructor mapping as a workaround.

**Comment by MarkUfland** [comment]
I have a workaround, that is a bit ugly, which involves allowing mapping of private set properties, removing the variable arguments out of the constructor and using the ForMember option with a custom resolver which takes me back to the solution at the top.

I will however, install VS 2017 and see if I can come up with a proper solution which should be to allow the use of a customer resolver in the ForCtorParam method.

**Comment by jbogard** [comment]
The free version of VS 2017 should work just fine to use if that's a problem.

**Comment by MarkUfland** [comment]
I have the Enterprise version I just had not got round to installing yet :-).

I have installed it now and gone through the project. What I have found in the tests is that you have already  implemented what I need in the form of

var mapper = new MapperConfiguration(
                cfg => cfg.CreateMap<Source, Dest>().ForCtorParam("thing", opt =>
                    opt.ResolveUsing((src, ctx) => ctx.Items[itemKey])
                ))
                .CreateMapper();

I have implemented it in my code and it worked fine. I'm not sure why I didn't see it. I probably didn't look hard enough at the signature in the intellisense but it was the weekend so that's my excuse lol.

Anyway, I'll close this off now.

Thanks
